### PR TITLE
Deleting user account

### PR DESCRIPTION
**Pull request by sanajaved7**
## Steps to test/reproduce
_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

## Show me
_Provide screenshots/animated gifs/videos if necessary._

# REMEMBER: Attach this PR to the Trello card

### CODE REVIEW

File: src/js/index.tsx [hunk 1]
```diff
@@ -92,6 +93,7 @@ const App = withRouter(
               <ErrorBoundary>
                 <Switch>
                   <Route exact path={routePatterns.login} component={Login} />
+                  <Route exact path={routePatterns.manage} component={Manage} />
```
Comment (jgerigmeyer) [inline src/js/index.tsx:96]: We probably want to use `<PrivateRoute>` here so that this page is only available when logged in.

```suggestion
                  <PrivateRoute exact path={routePatterns.manage} component={Manage} />
```

**Review by jgerigmeyer — COMMENTED** [review]
@sanajaved7 Looks great! I left some minor comments, but this is definitely on the right track. 🎉

File: metecho/api/models.py [hunk 2]
```diff
@@ -1198,7 +1206,7 @@ class ScratchOrg(
     description = MarkdownField(blank=True, property_suffix="_markdown")
     org_type = StringField(choices=ScratchOrgType.choices)
     org_config_name = StringField()
-    owner = models.ForeignKey(User, on_delete=models.PROTECT)
+    owner = models.ForeignKey(User, on_delete=models.PROTECT, blank=True, null=True)
```
Comment (jgerigmeyer) [inline metecho/api/models.py:1209]: Does this require changes to the `save` method below, where we reference `self.owner.id`?

Comment (jgerigmeyer) [inline metecho/api/models.py:1209]: This requires changes to `orgs/reducer.ts` where we define the Org `owner: string` (should now be `owner: string | null`).

File: src/js/store/tasks/reducer.ts [hunk 3]
```diff
@@ -156,6 +156,11 @@ const reducer = (
             },
           },
         };
+      } else if (objectType === OBJECT_TYPES.TASK && assigned_to_me) {
+        return {
+          tasks: response,
+          fetched: true,
+        };
```
Comment (jgerigmeyer) [inline src/js/store/tasks/reducer.ts:163]: This won't work quite right, because it's expecting the returned object here to be a `TaskState` object, which has keys as product IDs, and values of `TaskByProjectState` objects. So like I said in Slack, I'd lean toward not using `fetchObjects` to get this list, and not using the reducer at all or storing these in the Redux store.

File: metecho/api/models.py [hunk 4]
```diff
@@ -216,6 +216,14 @@ def _get_org_property(self, key):
         except (AttributeError, KeyError, TypeError):
             return None
 
+    def delete(self, *args, **kwargs):
+        for scratch_org_to_delete in self.scratchorg_set.all():
+            scratch_org_to_delete.owner = None
```
[the remark this answers is not in the document]
Comment (sanajaved7) [reply metecho/api/models.py:221]: @jerivas correct me if I'm wrong but we only had to set owner to None here in order to delete the scratch org. I think the other fields wouldn't block the scratch org from deletion, right?

[the remark this answers is not in the document]
Comment (jerivas) [reply metecho/api/models.py:221]: I wasn't aware of those fields. I agree we should set to None if we have lost the reference to the User

File: metecho/api/models.py (hunk 2 above)
Comment (jerivas) [reply metecho/api/models.py:1209]: I think we can modify change the [conditional](https://github.com/oddbird/Metecho/blob/658e2d5dceb1f3030c26d88e9a25e53a9adbea39/metecho/api/models.py#L1276) to `if is_new and self.owner:`

File: metecho/api/tests/models.py [hunk 5]
```diff
@@ -777,6 +777,23 @@ def test_is_devhub_enabled__sf_error(self, user_factory, social_account_factory)
             get_devhub_api.return_value = client
             assert not user.is_devhub_enabled
 
+    def test_user_delete_triggers_scratch_org_delete(
+        self, user_factory, scratch_org_factory
+    ):
+        user = user_factory()
+        with ExitStack() as stack:
```
Comment (jerivas) [inline metecho/api/tests/models.py:784]: `pytest-mock` provides a `mocker` fixture that replaces `ExitStack` (and the need for nesting in general). For new tests I would recommend we use [this approach](https://github.com/oddbird/Metecho/blob/658e2d5dceb1f3030c26d88e9a25e53a9adbea39/metecho/api/tests/admin.py#L101-L119) instead. You can find more examples by searching for `mocker.` in the tests.

File: metecho/api/tests/models.py [hunk 6]
```diff
@@ -842,6 +859,18 @@ def test_notify_delete(self, scratch_org_factory):
 
             assert async_to_sync.called
 
+    def test_delete_scratch_org_with_no_owner(self, scratch_org_factory):
```
Comment (jerivas) [inline metecho/api/tests/models.py:862]: I think we don't need this test, the case when `user=None` is already handled by the test on `User.delete()`

File: metecho/api/tests/views.py [hunk 7]
```diff
@@ -246,6 +246,10 @@ def test_check_app_installation__no_permissions(
             len(data["messages"]) == 3
         ), "Expected three error messages when permission checks fail"
 
+    def test_delete(self, client, mocker):
+        response = client.delete(reverse("current-user-detail"))
+        assert response.status_code == 204
```
Comment (jerivas) [inline metecho/api/tests/views.py:251]: Might be a good idea to test for the User not existing anymore at the end:

```python
with pytest.raises(client.user.DoesNotExist):
    client.user.refresh_from_db()
```

**Review by jerivas — CHANGES_REQUESTED** [review]
Looking good! Left some comments mostly on tests

File: metecho/api/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply metecho/api/models.py:221]: I think there also might be an existing bug where we [reassign a scratch org](https://github.com/oddbird/Metecho/blob/main/metecho/api/jobs.py#L1143), and we're neglecting to update those [additional fields](https://github.com/oddbird/Metecho/blob/main/metecho/api/jobs.py#L380-L382). 🤷

File: src/js/components/user/info.tsx [hunk 8]
```diff
@@ -285,6 +286,9 @@ const UserDropdown = () => {
                   >
                     {user.username}
                   </h2>
+                  <div className="slds-m-top_xx-small">
+                    <ManageAccountButton />
+                  </div>
```
Comment (jgerigmeyer) [inline src/js/components/user/info.tsx:291]: Hrm, functionally it's kind of annoying that clicking this link takes you to the Manage Account page, but does not automatically close this Popover. 🤔 We may want to make this a controlled component (using `onClick`, `onRequestClose`, and `isOpen` props) so that we can manually close the popover when this link is clicked.

File: src/js/utils/api.ts [hunk 9]
```diff
@@ -124,4 +126,20 @@ export const removeUrlParam = (key: string, search?: string) => {
...
+  if (userDeleteUrl) {
+    return userDeleteUrl;
+  }
+  const urlFn = window.api_urls[`${objectType}_detail`];
+  let baseUrl;
+  if (urlFn && id) {
+    baseUrl = urlFn(id);
+  }
+  return baseUrl;
+};
```
Comment (jgerigmeyer) [inline src/js/utils/api.ts:142]: Personally I still think it would be worth getting rid of this and replacing it with a more generic `url` option in the `deleteObject` action, which could be set to `window.api_urls.current_user_detail()` in the one case where we want that.

File: metecho/api/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jerivas) [reply metecho/api/models.py:221]: Created a follow-up card

File: src/js/components/utils/deleteModal.tsx [hunk 10]
```diff
@@ -10,16 +10,25 @@ import { ThunkDispatch } from '@/js/store';
...
+  model: Epic | Task | User;
   modelType: ObjectTypes;
   isOpen: boolean;
   redirect: string;
   handleClose: () => void;
 }
 
+const getModelDeleteUrl = (modelType: ObjectTypes) => {
+  let deleteUrl;
+  if (modelType === 'user') {
```
Comment (jgerigmeyer) [inline src/js/components/utils/deleteModal.tsx:26]:
```suggestion
  if (modelType === OBJECT_TYPES.USER) {
```

File: src/js/components/utils/deleteModal.tsx [hunk 11]
```diff
@@ -39,6 +48,7 @@ const DeleteModal = ({
       deleteObject({
         objectType: modelType,
         object: model,
+        url: getModelDeleteUrl(modelType),
```
Comment (jgerigmeyer) [inline src/js/components/utils/deleteModal.tsx:51]: I could go either way on whether the helper fn is useful here:

```suggestion
        url: modelType === OBJECT_TYPES.USER ? window.api_urls.current_user_detail() : undefined,
```

File: src/js/store/actions.ts [hunk 12]
```diff
@@ -329,14 +335,20 @@ export const deleteObject =
           id: object.id,
         });
       }
+      if (objectType === 'user') {
```
Comment (jgerigmeyer) [inline src/js/store/actions.ts:338]:
```suggestion
      if (objectType === OBJECT_TYPES.USER) {
```

**Review by jgerigmeyer — APPROVED** [review]
